### ISSUE TITLE
Check Yan algorithm on only expert classifications (small T)

### ISSUE THREAD

**Issue by MatthewJA** [chengsoonong/crowdastro#126]

**Comment by MatthewJA** [comment]
Only one expert annotator has classified anything. I'll take the people who've classified the most and run it on them.

**Comment by MatthewJA** [comment]
Took the top 10 most prolific annotators (each have an accuracy of ~80%) and ran everything through the passive learning, fully observed algorithm, downsampling negative examples. What does the fully observed algorithm do if the labels aren't fully observed? I have no idea, so I'm going to actually implement the partially observed algorithm now and retry this.

Nevertheless: 60.0% accuracy (worse than logistic regression on majority vote) with confusion matrix

```
[[21586   297]
 [ 1781   476]]
```

Weirdly enough, γ didn't change from the initial conditions. I can't make heads or tails of w.

**Comment by chengsoonong** [comment]
To analyse this better, look at the plot of the class conditional density. We discussed this on Monday on the bottom left of the whiteboard. The easiest way to do this is to plot the histograms of the scores on the training and test set (two plots). Two histograms on the same plot, one for each class. If the number of examples per class are vastly different, you may need to do some fancy two axis scaling on the y-axis (one for each class).

**Comment by MatthewJA** [comment]
Ah, right. Assuming I understood correctly, I should density plot the raw outputs for each true class, so I'll do that now.

**Comment by MatthewJA** [comment]
As an aside, `numpy.ma` is _really_ slow. Everything scales pretty poorly with basically everything: number of annotators, number of samples, and dimensionality of the data. Reducing the dimensionality of the data (through a CNN maybe?) seems to be the best bet for making things more tractable; maybe #137 will help.

**Comment by MatthewJA** [comment]
This is a great way of visualising the result!

![image](https://cloud.githubusercontent.com/assets/2681312/17540804/c542f7bc-5efc-11e6-955f-099254f16033.png)

Green plot is the density of positive examples (i.e. AGNs according to Norris) and red is the negatives (everything else). It's immediately clear that there are many false negatives. I find it really weird that there's two peaks for the positive examples &mdash; is this something I ought to expect from logistic regression, or is there likely a genuine difference between the false negatives and true positives? (One easy way this could be the case is that the right half is compact sources.)

**Comment by chengsoonong** [comment]
Two separate plots like the above:
1. Training set (predict on training set!)
2. Test set

**Comment by chengsoonong** [comment]
Masked arrays seem to be generally found to be slow (stack overflow has a few threads about this). It seems that masked arrays essentially do a 1D copy and concatenate (in python). You may have to manually keep a boolean mask around.
http://ipython-books.github.io/featured-01/

**Comment by MatthewJA** [comment]
I'm surprised that a manual boolean mask would be faster than NumPy &mdash; I would have thought NumPy would essentially do that anyway under the hood. This would also make dot products and broadcasted operations a lot harder since I'd have to unroll many of them (I think).

**Comment by chengsoonong** [comment]
It is a problem that still has not found consensus.
http://www.numpy.org/NA-overview.html

**Comment by MatthewJA** [comment]
> Two separate plots like the above:
> 1. Training set (predict on training set!)
> 2. Test set

Test set:
![image](https://cloud.githubusercontent.com/assets/2681312/17541171/250f5b38-5f00-11e6-9c8c-5740fdb454ec.png)

Train set:
![image](https://cloud.githubusercontent.com/assets/2681312/17541182/3ed45c3a-5f00-11e6-9710-61015059dac1.png)

**Comment by MatthewJA** [comment]
> It is a problem that still has not found consensus.
> http://www.numpy.org/NA-overview.html

Interesting read. It does seem that `ma` does the thing we need it to do, but also maybe masked/NA arrays are the wrong data structure for something this sparse. We may genuinely get performance improvement by taking out the arrays entirely. If worst came to worst, I could even go ahead and implement this with `for` loops in C or Rust and import it as a compiled Python module. The core algorithm is extremely simple so this wouldn't be too hard.
